### ISSUE TITLE
Keyboard navigation in menu

### ISSUE THREAD

**Issue by msxfm** [RequestPolicyContinued/requestpolicy#357]
<a href="https://github.com/umspokbl"><img src="https://avatars.githubusercontent.com/u/3305940?" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [umspokbl](https://github.com/umspokbl)**
_Friday Jan 18, 2013 at 12:36 GMT_
_Originally opened as https://github.com/RequestPolicy/requestpolicy/issues/357_

---

It would be a great enhancement to usability to be able to access the Permissions-Menu via keyboard shortcut (e.g., be able to tap CTRL+SHIFT+<some letter>) and have the Permissions-Menu open and be navigable via arrow keys. Users would be able to tap Enter to make selections and/or Esc to close the Permissions-Menu; users would also be able to tap alphabetical keys that correspond to "underlined-letters in Permissions-Menu descriptions" to directly access that selection.

**Comment by jstetten** [comment]
As a Vim/Pentadactyl/AutoHotKey user, I'm all about the keyboard shortcuts as well. NoScript's solution is really amazing in this regard. RPC's UI is completely different though, so I don't know yet what is needed to make this a reality. Even just arrow keys + Enter to navigate items would be huge.

**Comment by myrdd** [comment]
I'm open to pull requests. Arrows+Enter navigation seems like a good first step.

**Comment by jstetten** [comment]
Still quite busy, but would love to help make my much-used RPC be keyboard-navigable. It does sound like a lot of work when I think of all the key triggers, menu highlighting, and state-tracking code that presumably must be added. I can't immediately find another similar addon with this functionality. Maybe because it's not easy.

If the UI were rewritten as an HTML5 popup, with each line an actionable href link, plain old Tab, Arrow and Enter keys could automatically navigate it (not to mention vimperator/pentadactyl). For another day, perhaps....

Finally, I take it these [old comments](https://github.com/RequestPolicyContinued/requestpolicy/issues/47#issuecomment-48131644) were referring to the old RP interface, right?

**Comment by myrdd** [comment]
> Still quite busy, but would love to help make my much-used RPC be keyboard-navigable.

That would be great! :)

> If the UI were rewritten as an HTML5 popup, with each line an actionable href link

Why is this necessary to make the menu keyboard-navigable? Keyboard navigation should be possible with XUL+JS as well.

> Finally, I take it these old comments were referring to the old RP interface, right?

They are referring the "redirection notification bar", which is independent from the menu-style. The bar looks like this:

![redirection notification with flag icon](https://cloud.githubusercontent.com/assets/4286829/12374787/4daeb440-bca7-11e5-9649-0cf80b780d80.png)

The "access keys" mentioned in the comment you've linked to are those underlined `A`, `D`, and `r` letters. Combined with `Alt` you press that button.

**Comment by jstetten** [comment]
> Why is this necessary to make the menu keyboard-navigable? Keyboard navigation should be possible with XUL+JS as well.

Definitely. I guess I threw the thought out there because _if_ the popup were html5, then `Tab` and `Enter` (not to mention [vimperator](https://upload.wikimedia.org/wikipedia/commons/thumb/9/92/Vimp2_hints.png/400px-Vimp2_hints.png)) could navigate the links "automatically". Just brainstorming before bed, really.

Anyway, that was before I saw that `menupopup` provides [keypress event monitors](https://developer.mozilla.org/en/docs/Mozilla/Tech/XUL/PopupGuide/PopupKeys) already, along with highlighting and submenus. What's not to like? If I had time, I would test this out, but afraid I can't quite yet.

> They are referring the "redirection notification bar", which is independent from the menu-style.

Got it. The old issues are hard to track down. I already use those shortcuts and love them very much.  :)

**Comment by jstetten** [comment]
> They are referring the "redirection notification bar", which is independent from the menu-style.

Hey, @myrdd, while on this subject: After clicking the `Allow` button, can you think of any way of removing the redirection step from the tab's URL history? Currently, after clicking `Allow`, viewing the page, and clicking the browser's `Back` button, I get what appears to be a page refresh. There is not even any redirection bar to tell me what happened. I need to press `Back` twice to get back to where I was originally. 

I know it's a small detail, but when going through a lot of links quickly, it does start to take a toll on my flow at least. ("One `Back` or two? What did I do before...?")

**Comment by myrdd** [comment]
@jstetten I can confirm this behavior for http location header redirects. Would you mind opening a new issue for that? Your description should be fine when you also mention "http location header".

It should be possible to fix this bug.

**Comment by jstetten** [comment]
[Created](https://github.com/RequestPolicyContinued/requestpolicy/issues/780).
